### PR TITLE
feat: upgrade function extension and extension bundle

### PR DESCRIPTION
**Pull request by pudkrong**
Azure functions runtime version 3 reached end of life on 22 Dec 2022. So upgrade to latest version, 4 and node version to 18

### CODE REVIEW

File: feature-runner/steps/device/connectDevice.ts [hunk 1]
```diff
@@ -29,20 +29,34 @@ export const connectDevice = async ({
...
-			fs.readFile(
-				path.join(process.cwd(), 'data', 'DigiCertTLSECCP384RootG5.crt.pem'),
-				'utf-8',
+	const [
+		deviceKey,
+		deviceCert,
+		intermediateCA,
+		digiCert,
+		baltimore,
+		microsoftRSARoot,
```
Comment (coderbyheart) [inline feature-runner/steps/device/connectDevice.ts]: Since we are not migrating devices, we should only need the new DigiCertGlobalRootG2.pem, not the other two.

Comment (pudkrong) [reply feature-runner/steps/device/connectDevice.ts]: > Since we are not migrating devices, we should only need the new DigiCertGlobalRootG2.pem, not the other two.

According to MS docs, they said `The Microsoft RSA Root Certificate Authority 2017 helps prevent future disruptions in case the DigiCert Global Root G2 is retired unexpectedly.`

What do you think?

Comment (coderbyheart) [reply feature-runner/steps/device/connectDevice.ts]: That's not relevant for the software-based device used in the end-to-end test runner.

File: .github/workflows/test-and-release.yaml [hunk 2]
```diff
@@ -30,6 +30,7 @@ jobs:
       APP_REG_CLIENT_ID: ${{ secrets.APP_REG_CLIENT_ID }}
       B2C_CLIENT_SECRET: ${{ secrets.B2C_CLIENT_SECRET }}
       B2C_TENANT_ID: ${{ secrets.B2C_TENANT_ID }}
+      B2C_USER_FLOW_NAME: B2C_1_developer
```
Comment (coderbyheart) [inline .github/workflows/test-and-release.yaml:33]: I assume this change is needed for CI to work.

File: .github/workflows/test-and-release.yaml [hunk 3]
```diff
@@ -110,11 +111,11 @@ jobs:
         with:
           inlineScript: |
             az keyvault secret set --vault-name ${keyVaultName} --name unwiredlabsApiKey --value my-secret
-            openssl ecparam -name secp256k1 -genkey > nrfcloudcelllocationservice.key
+            openssl ecparam -name prime256v1 -genkey > nrfcloudcelllocationservice.key
```
Comment (coderbyheart) [inline .github/workflows/test-and-release.yaml:114]: Unrelated change.

File: data/BaltimoreCyberTrustRoot.pem [hunk 4]
```diff
@@ -18,4 +18,4 @@ jkzSswF07r51XgdIGn9w/xZchMB5hbgF/X++ZRGjD8ACtPhSNzkE1akxehi/oCr0
 Epn3o0WC4zxe9Z2etciefC7IpJ5OCBRLbf1wbWsaY71k5h+3zvDyny67G7fyUIhz
 ksLi4xaNmjICq44Y3ekQEe5+NauQrz4wlHrQMz2nZQ/1/I6eYs9HRCwBXbsdtTLS
 R9I4LtD+gdwyah617jzV/OeBHRnDJELqYzmp
------END CERTIFICATE-----
\ No newline at end of file
+-----END CERTIFICATE-----
```
Comment (coderbyheart) [inline data/BaltimoreCyberTrustRoot.pem]: Unrelated change.

File: data/DigiCertGlobalRootG2.pem
Comment (coderbyheart) [inline data/DigiCertGlobalRootG2.pem:1]: I assume this is needed for CI to work.

File: geolocateCellFromNrfCloud/geolocateCell.ts [hunk 5]
```diff
@@ -142,6 +143,7 @@ const geolocateCell: AzureFunction = async (
 					404,
 				)
 				context.bindings.cellGeolocation = JSON.stringify({
+					id: randomUUID(),
```
Comment (coderbyheart) [inline geolocateCellFromNrfCloud/geolocateCell.ts]: Use cellID as id.

File: mock-http-api.json [hunk 6]
```diff
@@ -75,15 +75,15 @@
...
+              "value": "~4"
             },
             {
               "name": "FUNCTIONS_WORKER_RUNTIME",
               "value": "node"
             },
             {
               "name": "WEBSITE_NODE_DEFAULT_VERSION",
-              "value": "~14"
+              "value": "~18"
```
Comment (coderbyheart) [inline mock-http-api.json:86]: Can we updated to Node 20?

File: package.json [hunk 7]
```diff
@@ -59,7 +59,7 @@
     "@types/chai-subset": "1.3.3",
     "@types/jest": "29.4.0",
     "@types/jsonwebtoken": "9.0.1",
-    "@types/node": "18.14.0",
+    "@types/node": "18.18.0",
```
Comment (coderbyheart) [inline package.json]: Unrelated change.

File: storeDeviceCellGeolocation/storeDeviceCellGeolocation.ts
Comment (coderbyheart) [inline storeDeviceCellGeolocation/storeDeviceCellGeolocation.ts:1]: Remove device cell geo location feature, see https://github.com/NordicSemiconductor/asset-tracker-cloud-aws-js/commit/1e0735453af47a2ebae0e6fecc88935e16244a7b

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: .github/workflows/test-and-release.yaml (hunk 2 above)
Comment (pudkrong) [reply .github/workflows/test-and-release.yaml:33]: Previously, it is hardcode. So, making it as parameter will give us more flexible to test somewhere else

File: .github/workflows/test-and-release.yaml (hunk 3 above)
Comment (pudkrong) [reply .github/workflows/test-and-release.yaml:114]: It is required to fix the error on https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/e0bc28ebb5a4d0dd04d2d5e66c8610c6d9f13324/third-party/nrfcloud.com/apiclient.ts#L422. 

According to jsonwebtoken lib requires curve 'prime256v1' when using 'ES256' algorithm,
https://github.com/auth0/node-jsonwebtoken/issues/862

File: data/DigiCertGlobalRootG2.pem
Comment (pudkrong) [reply data/DigiCertGlobalRootG2.pem:1]: yes, it is required to enable device simulator to connect to IoT hub

File: mock-http-api.json (hunk 6 above)
Comment (pudkrong) [reply mock-http-api.json:86]: It is still in public preview, https://azure.microsoft.com/en-us/updates/public-preview-azure-functions-support-for-nodejs-20/#:~:text=Published%20date%3A%20September%2027%2C%202023&text=You%20can%20now%20develop%20functions,your%20Function%20Apps%20to%20Node. So, do you still want to use node20?

Comment (pudkrong) [reply mock-http-api.json:86]: Another thing is, if we use node20, there is warning error from `@azure/msal-node@1.24.2` as

```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@azure/msal-node@1.14.2',
npm WARN EBADENGINE   required: { node: '10 || 12 || 14 || 16 || 18' },
npm WARN EBADENGINE   current: { node: 'v20.8.0', npm: '9.7.2' }
npm WARN EBADENGINE }
```

File: package.json (hunk 7 above)
Comment (pudkrong) [reply package.json]: Using version `18.14.0`, I have found error about type mismatch on `node-fetch` at https://github.com/NordicSemiconductor/asset-tracker-cloud-azure-js/blob/e0bc28ebb5a4d0dd04d2d5e66c8610c6d9f13324/feature-runner/lib/b2c/accessToken.ts#L58

The error message is

```
Type 'URLSearchParams' is not assignable to type 'BodyInit'.
  Property 'size' is missing in type 'import("url").URLSearchParams' but required in type 'URLSearchParams'.ts(2322)
lib.dom.d.ts(22575, 14): 'size' is declared here.
index.d.ts(75, 2): The expected type comes from property 'body' which is declared here on type 'RequestInit'
```

File: .github/workflows/test-and-release.yaml (hunk 3 above)
Comment (coderbyheart) [reply .github/workflows/test-and-release.yaml:114]: So, it's unrelated but required for CI to pass?

File: mock-http-api.json (hunk 6 above)
Comment (coderbyheart) [reply mock-http-api.json:86]: Yes, because it is LTS in 12 days: https://github.com/nodejs/release#release-schedule

File: package.json (hunk 7 above)
Comment (coderbyheart) [reply package.json]: Get rid of node-fetch and use native fetch (which is available since v18)

File: .github/workflows/test-and-release.yaml (hunk 3 above)
Comment (pudkrong) [reply .github/workflows/test-and-release.yaml:114]: Yes

File: package.json (hunk 7 above)
Comment (pudkrong) [reply package.json]: After removing `node-fetch` and use native fetch, I can then revert `@types/node`changes.

File: .github/workflows/test-and-release.yaml (hunk 2 above)
Comment (pudkrong) [reply .github/workflows/test-and-release.yaml:33]: This change is needed for CI to work

File: mock-http-api.json (hunk 6 above)
Comment (pudkrong) [reply mock-http-api.json:86]: To upgrade to Node v20, Azure function has to migrate to programming model v4 which has a big change. So, we will use Node v18 for now. It will be another PR to upgrade to Node v20.

File: geolocateCellFromNrfCloud/geolocateCell.ts [hunk 8]
```diff
@@ -142,6 +142,7 @@ const geolocateCell: AzureFunction = async (
 					404,
 				)
 				context.bindings.cellGeolocation = JSON.stringify({
+					id,
 					cellId: id,
```
Comment (coderbyheart) [inline geolocateCellFromNrfCloud/geolocateCell.ts]: do not duplicate the value.

**Review by coderbyheart — CHANGES_REQUESTED** [review]

**Review by coderbyheart — APPROVED** [review]

**Comment by coderbyheart** [comment]
Great!
